### PR TITLE
Mirko/fortress self hosted

### PR DESCRIPTION
**Pull request by MirkoFerrati**

### CODE REVIEW

**Review by artivis — DISMISSED** [review]
Approving for testing only, no merging :P

File: .github/workflows/snap.yaml [hunk 1]
```diff
@@ -12,7 +12,7 @@ on:
 
 jobs:
   build:
-    runs-on: ubuntu-latest
+    runs-on: [self-hosted, linux, X64, jammy, xlarge]
```
Comment (Guillaumebeuzeboc) [inline .github/workflows/snap.yaml]: What is all that? why not just using ubuntu-latest?

File: snap/snapcraft.yaml [hunk 2]
```diff
@@ -75,8 +74,13 @@ parts:
...
       - "-DCMAKE_PREFIX_PATH=$(echo $SNAPCRAFT_CMAKE_ARGS | awk -F= '{printf(\"%s/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/cmake/Qt5\", $2)}')"
       - "--event-handlers=console_direct+"
+      - "--executor sequential"
+
     build-environment:
-      - IGNITION_VERSION: citadel
+      - IGNITION_VERSION: fortress
+      - GZ_VERSION: fortress
+      - PKG_CONFIG_PATH: ""
+      - MAKEFLAGS: " -j1"
```
Comment (Guillaumebeuzeboc) [inline snap/snapcraft.yaml]: Why do you specify sequential and only one thread?

File: snap/snapcraft.yaml [hunk 3]
```diff
@@ -295,28 +209,46 @@ apps:
...
-    extensions: [ros2-foxy]
+    extensions: [ros2-humble, kde-neon]
 
   ros-gz-bridge:
-    command: opt/ros/foxy/bin/ros2 run ros_ign_bridge
+    command: ros2 run ros_gz_bridge
     environment:
-        FASTRTPS_DEFAULT_PROFILES_FILE: ${SNAP}/usr/share/fastdds_no_shared_memory.xml
+      FASTRTPS_DEFAULT_PROFILES_FILE: ${SNAP}/usr/share/fastdds_no_shared_memory.xml
+      LD_LIBRARY_PATH: '$SNAP/usr/local/lib/:$SNAP/usr/local/lib/$SNAPCRAFT_ARCH_TRIPLET/:$SNAP/usr/lib/:$SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/:$SNAP/usr/local/lib/site_ruby/3.0.0:$SNAP/usr/local/lib/$SNAPCRAFT_ARCH_TRIPLET/site_ruby:$SNAP/usr/local/lib/site_ruby:$SNAP/usr/lib/ruby/vendor_ruby/3.0.0:$SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/ruby/vendor_ruby/3.0.0:$SNAP/usr/lib/ruby/vendor_ruby:$SNAP/usr/lib/ruby/3.0.0:$SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/ruby/3.0.0:${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}'
```
Comment (Guillaumebeuzeboc) [inline snap/snapcraft.yaml]: This could be set in a command-chain to make it cleaner

File: snap/local/cleanup.sh [hunk 4]
```diff
@@ -0,0 +1,66 @@
+#!/bin/bash
```
Comment (Guillaumebeuzeboc) [inline snap/local/cleanup.sh:1]: why not reusing https://github.com/ubuntu-robotics/snapcraft-ros-clean directly?

File: snap/snapcraft.yaml [hunk 5]
```diff
@@ -75,8 +74,13 @@ parts:
       - "-DCMAKE_SKIP_RPATH=TRUE"
       - "-DCMAKE_PREFIX_PATH=$(echo $SNAPCRAFT_CMAKE_ARGS | awk -F= '{printf(\"%s/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/cmake/Qt5\", $2)}')"
       - "--event-handlers=console_direct+"
+      - "--executor sequential"
```
Comment (artivis) [inline snap/snapcraft.yaml]: Why `sequential`?

File: snap/snapcraft.yaml [hunk 6]
```diff
@@ -75,8 +74,13 @@ parts:
...
+
     build-environment:
-      - IGNITION_VERSION: citadel
+      - IGNITION_VERSION: fortress
+      - GZ_VERSION: fortress
+      - PKG_CONFIG_PATH: ""
+      - MAKEFLAGS: " -j8"
```
Comment (artivis) [inline snap/snapcraft.yaml]: I think this is already being taken care of [here](https://github.com/snapcore/snapcraft/blob/9ad7a93ea880b560d3c8a364617c67a3b0b6d157/snapcraft/parts/plugins/colcon_plugin.py#L184).

File: snap/snapcraft.yaml (hunk 5 above)
Comment (MirkoFerrati) [reply snap/snapcraft.yaml]: Because I needed determinism when debugging the building issue with standard GitHub runners. And I kinda liked it since I can compare logs and timings. I don't think it hurts the speed that much given how the codebase is structured. After we merge I can try to remove it and compare the timings

File: snap/snapcraft.yaml [hunk 7]
```diff
@@ -295,28 +209,46 @@ apps:
...
+      LD_LIBRARY_PATH: '$SNAP/usr/local/lib/:$SNAP/usr/local/lib/$SNAPCRAFT_ARCH_TRIPLET/:$SNAP/usr/lib/:$SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/:$SNAP/usr/local/lib/site_ruby/3.0.0:$SNAP/usr/local/lib/$SNAPCRAFT_ARCH_TRIPLET/site_ruby:$SNAP/usr/local/lib/site_ruby:$SNAP/usr/lib/ruby/vendor_ruby/3.0.0:$SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/ruby/vendor_ruby/3.0.0:$SNAP/usr/lib/ruby/vendor_ruby:$SNAP/usr/lib/ruby/3.0.0:$SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/ruby/3.0.0:${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}'
+      IGN_GAZEBO_RENDER_ENGINE_PATH: '$SNAP/opt/ros/snap/lib/ign-rendering-6/engine-plugins/'
+      GZ_SIM_RENDER_ENGINE_PATH: '$SNAP/opt/ros/snap/lib/ign-rendering-6/engine-plugins/'
     command: usr/bin/ruby $SNAP/opt/ros/snap/bin/ign
-    plugs: [network, network-bind, home]
-    extensions: [kde-neon, ros2-foxy]
+    plugs:
+    - network
+    - network-bind
+    - home
```
Comment (artivis) [inline snap/snapcraft.yaml]: Fix indentation please.

File: scripts/override_pull.sh [hunk 8]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
```
Comment (artivis) [inline scripts/override_pull.sh]: I think this script should be in `snap/local` really.

File: snap/snapcraft.yaml (hunk 5 above)
Comment (artivis) [reply snap/snapcraft.yaml]: Gotcha.
Indeed it would be nice to see how much time can be saved having the build be parallel.

File: snap/snapcraft.yaml [hunk 9]
```diff
@@ -220,73 +194,11 @@ parts:
       'fastdds_no_shared_memory.xml': usr/share/
 
   cleanup:
-    after: [gazebo, ros2-foxy-extension]
+    after: [gazebo, ros2-humble]
```
Comment (MirkoFerrati) [inline snap/snapcraft.yaml]: @Guillaumebeuzeboc  this passed with ros2-humble-extension on the first run, subsequent runs failed unless using ros2-humble as it is here. do we have an issue in our extension naming?

File: snap/snapcraft.yaml (hunk 3 above)
Comment (MirkoFerrati) [reply snap/snapcraft.yaml]: done

File: snap/local/cleanup.sh (hunk 4 above)
Comment (MirkoFerrati) [reply snap/local/cleanup.sh:1]: it's a single file, I prefer to avoid git pulling an entire repo just to do some cleanup here

File: snap/snapcraft.yaml (hunk 5 above)
Comment (MirkoFerrati) [reply snap/snapcraft.yaml]: removed

File: scripts/override_pull.sh (hunk 8 above)
Comment (MirkoFerrati) [reply scripts/override_pull.sh]: moved

File: snap/snapcraft.yaml (hunk 6 above)
Comment (MirkoFerrati) [reply snap/snapcraft.yaml]: removed

File: snap/snapcraft.yaml (hunk 9 above)
Comment (Guillaumebeuzeboc) [reply snap/snapcraft.yaml]: `ros2-humble-extension` name doesn't exist. Would you have the link of the CI that passed with this name?

Comment (MirkoFerrati) [reply snap/snapcraft.yaml]: any green run before the change above, for example https://github.com/canonical/gazebo_snap/actions/runs/6025428362/job/16346109543 yesterday

Comment (Guillaumebeuzeboc) [reply snap/snapcraft.yaml]: The commit ref: d3e0df9f4c14218699a25a91a97036198525364b cannot be found in your branch. Where can I find this commit?

Comment (Guillaumebeuzeboc) [reply snap/snapcraft.yaml]: So the comment shows the commit ^^
and you are using the `ros2-humble` extension not the `ros2-humble-extension` extension
https://github.com/canonical/gazebo_snap/commit/d3e0df9f4c14218699a25a91a97036198525364b#diff-56759910381a014fecfd7556dd72ddd68c747d922a5b7df2044b9ce7c552f5f5R246

Comment (artivis) [reply snap/snapcraft.yaml]: Note that this lists other parts for which the `cleanup` must wait. It is not related to the extension name. Now the extension `ros-humble` do create its own part to dump the command-chain scripts. I don't recall the exact name of that part at the moment but I think you can omit it here.

File: .github/workflows/snap.yaml [hunk 10]
```diff
@@ -12,19 +12,19 @@ on:
...
+    - uses: snapcore/action-build@master
       id: build-snap
       with:
-        snapcraft-args: '--enable-experimental-extensions'
+        snapcraft-args: '--enable-experimental-extensions --verbose'
 
     # Make sure the snap is installable
     - run: |
-        sudo snap install --dangerous ${{ steps.build-snap.outputs.snap }}
+        sudo snap install --devmode --dangerous ${{ steps.build-snap.outputs.snap }}
```
Comment (artivis) [inline .github/workflows/snap.yaml:27]: Remind me why installing it in `devmode`?

File: snap/snapcraft.yaml (hunk 9 above)
Comment (artivis) [reply snap/snapcraft.yaml]: You don't have a `ros2-humble` part in this file, you should remove it. Not sure why it isn't complaining about that.  
Related to my previous comment in this thread, the Humble extension introduces a part that's named `ros2-humble/ros2-launch`. It was until core20 that there was a `ros2-foxy-extension` part schema.

File: snap/snapcraft.yaml [hunk 11]
```diff
@@ -295,28 +207,46 @@ apps:
 
   gz:
     environment:
-      RUBYLIB: '$SNAP/usr/local/lib/site_ruby/2.7.0:$SNAP/usr/local/lib/$SNAPCRAFT_ARCH_TRIPLET/site_ruby:$SNAP/usr/local/lib/site_ruby:$SNAP/usr/lib/ruby/vendor_ruby/2.7.0:$SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/ruby/vendor_ruby/2.7.0:$SNAP/usr/lib/ruby/vendor_ruby:$SNAP/usr/lib/ruby/2.7.0:$SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/ruby/2.7.0'
+      RUBYLIB: '$SNAP/usr/local/lib/site_ruby/3.0.0:$SNAP/usr/local/lib/$SNAPCRAFT_ARCH_TRIPLET/site_ruby:$SNAP/usr/local/lib/site_ruby:$SNAP/usr/lib/ruby/vendor_ruby/3.0.0:$SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/ruby/vendor_ruby/3.0.0:$SNAP/usr/lib/ruby/vendor_ruby:$SNAP/usr/lib/ruby/3.0.0:$SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/ruby/3.0.0'
       IGN_HOMEDIR: '$SNAP_USER_DATA'
+      GZ_HOMEDIR: '$SNAP_USER_DATA'
```
Comment (artivis) [inline snap/snapcraft.yaml:213]: There is a lot more GZ/IGN environnent variables defined. Are they all necessary? Can multiple paths be defined in those variables? Asking in case we need them one day for finding ressources outside this snap (e.g. content-sharing).

File: .github/workflows/snap.yaml [hunk 12]
```diff
@@ -12,19 +12,19 @@ on:
 
 jobs:
   build:
-    runs-on: ubuntu-latest
+    runs-on: [self-hosted, linux, X64, jammy, two-xlarge] # test on self hosted canonical runners
```
Comment (Guillaumebeuzeboc) [inline .github/workflows/snap.yaml]: Maybe no need for the comment "test"

File: .github/workflows/snap.yaml [hunk 13]
```diff
@@ -12,19 +12,19 @@ on:
...
     outputs:
       snap-file: ${{ steps.build-snap.outputs.snap }}
     steps:
     - uses: actions/checkout@v2
-    - uses: snapcore/action-build@v1
+    - uses: snapcore/action-build@master
```
Comment (Guillaumebeuzeboc) [inline .github/workflows/snap.yaml]: I would rather use a version in case the API breaks

File: .github/workflows/snap.yaml [hunk 14]
```diff
@@ -12,19 +12,19 @@ on:
...
       id: build-snap
       with:
-        snapcraft-args: '--enable-experimental-extensions'
+        snapcraft-args: '--enable-experimental-extensions --verbose'
```
Comment (Guillaumebeuzeboc) [inline .github/workflows/snap.yaml]: `ros2-humble` is not an experimental extension anymore, you can remove it.

File: .github/workflows/snap.yaml [hunk 15]
```diff
@@ -35,8 +35,12 @@ jobs:
         name: gazebo-snap
         path: ${{ steps.build-snap.outputs.snap }}
 
+    - name: "Print craft logs"
+      if: always()
```
Comment (Guillaumebeuzeboc) [inline .github/workflows/snap.yaml:39]: Why do you need `if: always()`?
Can't you just remove the line?

File: snap/snapcraft.yaml [hunk 16]
```diff
@@ -60,7 +60,6 @@ package-repositories:
 
 parts:
   gazebo:
-    after: [kde-neon-extension]
```
Comment (Guillaumebeuzeboc) [inline snap/snapcraft.yaml:63]: Why did you remove this if you need the kde-neon sdk part to build this one?

File: snap/snapcraft.yaml (hunk 9 above)
Comment (Guillaumebeuzeboc) [reply snap/snapcraft.yaml]: Unfortunately, in case of a typo snapcraft doesn't raise an error: https://bugs.launchpad.net/snapcraft/+bug/2033882
If the after keyword with the exact extension part name is not mentioned, there is no guarantee that the clean-up will happen after everything has been moved to `prime/`

**Review by Guillaumebeuzeboc — CHANGES_REQUESTED** [review]

File: snap/snapcraft.yaml (hunk 9 above)
Comment (MirkoFerrati) [reply snap/snapcraft.yaml]: This is tricky, how will our users know to cleanup after "ros2-humble/ros2-launch" ? this string is not going to show easily to new users

File: .github/workflows/snap.yaml (hunk 10 above)
Comment (MirkoFerrati) [reply .github/workflows/snap.yaml:27]: because our LXD containers are not setup with a user and without devmode the snap will fail to run since it looks for a user. See here https://forum.snapcraft.io/t/cannot-launch-snap-applications-with-cgroup-v2/27700/2 and in our case the env var XDG_RUNTIME_DIR is empty so quite hard to fix without devmode.

File: .github/workflows/snap.yaml (hunk 15 above)
Comment (MirkoFerrati) [reply .github/workflows/snap.yaml:39]: github syntax, this is how you tell github to run this even if there are errors. Otherwise the default is "run if success" when the line is removed and the other option is "run if error".

File: snap/snapcraft.yaml (hunk 16 above)
Comment (MirkoFerrati) [reply snap/snapcraft.yaml:63]: because it should be "kde-neon/sdk" and again I didn't know how to fetch the name, I have to fix this

File: snap/snapcraft.yaml (hunk 11 above)
Comment (MirkoFerrati) [reply snap/snapcraft.yaml:213]: I have been adding each of them after getting an error. What is not necessary is the duplication betwen GZ and IGN_GAZEBO. Will remove as soon as the next gazebo version is out with no deprecated env-vars anymore. 
No, multiple paths do not seem to be supported (I briefly checked the code for a couple of those) but the portability effort is going good so in a future iteration I expect to remove most of these env-vars including the ones that are intended for the user and not for the installation folder.

File: snap/snapcraft.yaml (hunk 9 above)
Comment (Guillaumebeuzeboc) [reply snap/snapcraft.yaml]: You mean when they write their own snap?

**Review by Guillaumebeuzeboc — APPROVED** [review]
